File: reference/numpy/cube_cavity.py
"""NumPy reference for the scalar-Helmholtz cube-cavity eigenmode pipeline.

Issue #92 (Epic #88, Phase B): mirrors the full Burn pipeline in NumPy
so the two backends can be cross-checked sub-stage by sub-stage.

The pipeline
============

1. Mesh I/O — either generate the canonical tet-split cube
   (``cube_tet_mesh(n)``) or load a ``.msh`` via ``meshio``
   (``load_msh(path)``). Both live in :mod:`mesh` (issue #103) and are
   re-exported here for backward compatibility with consumers like
   ``gen_cube_cavity_baseline.py``.
2. P1 local matrices — delegated to :mod:`p1_local_matrices` (NumPy
   reference shared with issue #90; pinned by the per-case fixtures
   under ``reference/fixtures/p1_local/<case>.json`` per #101).
3. Global assembly — scatter per-element ``[4, 4]`` local matrices into
   global ``[n_nodes, n_nodes]`` CSR via
   ``scipy.sparse.coo_matrix(...).tocsr()``. Stays close to the math:
   no clever optimization, no symmetry compression.
4. Dirichlet boundary conditions — restrict K, M to the interior nodes
   (``cube_interior_mask``) by row+column extraction.
5. Generalized eigensolve — ``scipy.sparse.linalg.eigsh`` with
   ``sigma=0`` shift-and-invert at the lowest end of the spectrum.

The eigenvalues are then in the same units as the Burn path. The
expected analytic targets for the unit cube with Dirichlet boundaries
are the Laplacian eigenvalues ``(p² + q² + r²)·π²`` for positive
integers ``p, q, r``:

    λ_0 = (1+1+1)·π² = 3π²        (mode (1,1,1))
    λ_1 = λ_2 = λ_3 = 6π²         (modes (2,1,1), (1,2,1), (1,1,2))
    λ_4 = 9π²                     (mode (2,2,1) permutations — 3 modes
                                   sharing this value, but on a coarse
                                   mesh only the first is recovered in
                                   the lowest 5 alongside the 3 modes
                                   at 6π². See README for details.)

On an n=10 mesh both the Burn path (issue #3) and this NumPy reference
hit the lowest 5 modes ``{3, 6, 6, 6, 9}·π²`` to roughly 4% relative
error — the residual being the standard P1 discretization error
``O(h²)`` on a 10×10×10 cube.

Public API
==========

Mesh primitives (re-exported from :mod:`mesh` for backward compat):

- :func:`cube_tet_mesh(n)` — generate the tet-split cube mesh.
- :func:`load_msh(path)` — read a Gmsh ``.msh`` file via ``meshio``.
- :func:`write_msh(path, nodes, tets)` — write a Gmsh ``.msh`` file.
- :func:`cube_interior_mask(nodes)` — boolean mask of interior nodes.

Assembly + eigensolve:

- :func:`assemble_global_p1(nodes, tets)` — assembled ``(K, M)`` as
  scipy CSR matrices.
- :func:`apply_dirichlet(K, M, mask)` — restrict K, M to interior DOFs.
- :func:`eigensolve(K, M, k=5)` — lowest-``k`` generalized eigenpairs.
- :func:`run_cube_cavity(n=10, k=5)` — orchestrator returning a dict
  with all intermediate quantities for cross-backend comparison.
"""

from __future__ import annotations

import inspect
import sys
from pathlib import Path

import numpy as np
import scipy.sparse
import scipy.sparse.linalg

# Allow `python3 cube_cavity.py` to find the sibling modules regardless of cwd.
sys.path.insert(0, str(Path(__file__).resolve().parent))
from p1_local_matrices import batched_p1_local_matrices  # noqa: E402

# Mesh primitives live in `mesh.py` (issue #103) and are re-exported
# here so existing callers (`gen_cube_cavity_baseline.py`, downstream
# tests) keep importing `cube_tet_mesh`, `load_msh`, `write_msh`, and
# `cube_interior_mask` from `cube_cavity` without churn.
from mesh import (  # noqa: E402, F401
    cube_interior_mask,
    cube_tet_mesh,
    load_msh,
    write_msh,
)


# --------------------------------------------------------------------------- #
# Global P1 assembly via coo_matrix.tocsr()
# --------------------------------------------------------------------------- #


def assemble_global_p1(nodes, tets):
    """Assemble global stiffness ``K`` and consistent mass ``M`` (CSR).

    Builds per-element ``[4, 4]`` locals (from
    :func:`p1_local_matrices.batched_p1_local_matrices`) then scatters
    into the global COO triplet, which scipy's COO->CSR conversion
    collapses duplicates by sum (the documented behavior of
    ``coo_matrix.tocsr``).

    Stays close to the math; no clever sparsity-pattern caching. The
    cube-cavity n=10 mesh is well under 10⁴ DOFs so wall time is a
    non-issue.

    Returns
    -------
    K : scipy.sparse.csr_matrix, shape ``(n_nodes, n_nodes)``, dtype f64
    M : scipy.sparse.csr_matrix, shape ``(n_nodes, n_nodes)``, dtype f64
    """
    nodes = np.asarray(nodes, dtype=np.float64)
    tets = np.asarray(tets, dtype=np.int64)
    n_nodes = nodes.shape[0]
    n_elem = tets.shape[0]

    # Per-element vertex coords: shape (n_elem, 4, 3).
    coords = nodes[tets, :]
    k_local, m_local, _signed_v = batched_p1_local_matrices(coords)

    # Row / col index arrays for every (e, i, j) triple, shape (n_elem*16,).
    # rows[e*16 + i*4 + j] = tets[e, i]
    # cols[e*16 + i*4 + j] = tets[e, j]
    rows = np.repeat(tets, 4, axis=1).reshape(n_elem * 16)
    cols = np.tile(tets, (1, 4)).reshape(n_elem * 16)

    k_vals = k_local.reshape(n_elem * 16)
    m_vals = m_local.reshape(n_elem * 16)

    K = scipy.sparse.coo_matrix(
        (k_vals, (rows, cols)), shape=(n_nodes, n_nodes)
    ).tocsr()
    M = scipy.sparse.coo_matrix(
        (m_vals, (rows, cols)), shape=(n_nodes, n_nodes)
    ).tocsr()

    return K, M


def apply_dirichlet(K, M, mask):
    """Restrict ``K, M`` to the rows/cols where ``mask`` is True.

    Returns ``(K_int, M_int)`` of shape ``(n_int, n_int)``. The dropped
    boundary rows/cols implement homogeneous Dirichlet conditions on
    the eliminated DOFs.
    """
    mask = np.asarray(mask, dtype=bool)
    interior = np.flatnonzero(mask)
    # CSR slicing the row dim then the col dim is the standard idiom.
    K_int = K[interior, :][:, interior]
    M_int = M[interior, :][:, interior]
    # Convert back to CSR (the chained slice yields CSR in scipy >= 1.4).
    return K_int.tocsr(), M_int.tocsr()


# --------------------------------------------------------------------------- #
# Generalized eigensolve
# --------------------------------------------------------------------------- #


def _deterministic_arpack_kwargs(n, solver, complex_pencil=False):
    """Deterministic ARPACK kwargs (issue #191).

    Fixed-seed normalized start vector ``v0`` so near-degenerate
    clusters converge reproducibly run-to-run; on scipy >= 1.17 also a
    fixed-seed ``rng``, because the rewritten ARPACK wrapper draws its
    internal *restart* vectors from OS entropy even when ``v0`` is
    pinned (older Fortran-ARPACK scipy is deterministic given ``v0``).
    """
    rng = np.random.default_rng(0)
    if complex_pencil:
        v0 = rng.standard_normal(n) + 1j * rng.standard_normal(n)
    else:
        v0 = rng.standard_normal(n)
    v0 = v0 / np.linalg.norm(v0)
    kwargs = {"v0": v0}
    if "rng" in inspect.signature(solver).parameters:
        kwargs["rng"] = np.random.default_rng(0)
    return kwargs


def eigensolve(K, M, k: int = 5):
    """Lowest-k generalized eigenpairs of ``K x = λ M x``.

    Uses ``scipy.sparse.linalg.eigsh`` (ARPACK-backed) with
    shift-and-invert at ``sigma=0`` and ``which='LM'`` — the canonical
    "smallest eigenvalues of a symmetric generalized pencil" recipe.

    Both ``K`` and ``M`` should be symmetric; ``M`` should be SPD.

    Returns
    -------
    eigvals : ndarray, shape ``(k,)``, dtype f64, ascending order.
    eigvecs : ndarray, shape ``(n_int, k)``, dtype f64.
        Columns are the corresponding eigenvectors, M-orthonormal
        (eigsh normalizes columns to satisfy ``v_i^T M v_j = δ_ij``).
    """
    # eigsh's shift-and-invert path requires a sparse SPD K-σM; at σ=0 that
    # is just K, which is positive *semi*-definite (the constant mode is in
    # the null space, but Dirichlet-restricted K is SPD on the interior).
    # Deterministic ARPACK iterations: reproducibility for
    # near-degenerate clusters (issue #191).
    det = _deterministic_arpack_kwargs(K.shape[0], scipy.sparse.linalg.eigsh)
    eigvals, eigvecs = scipy.sparse.linalg.eigsh(
        K.astype(np.float64),
        k=k,
        M=M.astype(np.float64),
        sigma=0.0,
        which="LM",
        **det,
    )
    # eigsh in shift-invert mode returns eigenvalues not necessarily sorted.
    order = np.argsort(eigvals)
    return eigvals[order], eigvecs[:, order]


# --------------------------------------------------------------------------- #
# End-to-end driver — used by the fixture generator and standalone runs
# --------------------------------------------------------------------------- #


def run_cube_cavity(
    n: int = 10,
    k: int = 5,
    mesh_path: str | None = None,
    side: float = 1.0,
):
    """Full cube-cavity pipeline; returns a dict for cross-backend comparison.

    Either ``n`` (generate the mesh in-process) or ``mesh_path`` (load a
    Gmsh ``.msh``) controls the mesh source.

    Returns
    -------
    dict with keys:
        - ``n_nodes``, ``n_tets``, ``n_int``: shape diagnostics
        - ``nodes``, ``tets``: the mesh used
        - ``interior_mask``: boolean array length ``n_nodes``
        - ``K``, ``M``: global CSR matrices (full, pre-Dirichlet)
        - ``K_int``, ``M_int``: interior CSR matrices (post-Dirichlet)
        - ``eigenvalues``: ascending, length ``k``
        - ``eigenvectors``: M-orthonormal columns, shape ``(n_int, k)``
        - ``k_frobenius``, ``m_frobenius``: Frobenius norms of K, M (full)
        - ``k_int_frobenius``, ``m_int_frobenius``: same for interior
        - ``k_int_diag``, ``m_int_diag``: diagonals of K_int, M_int
    """
    if mesh_path is not None:
        nodes, tets = load_msh(mesh_path)
    else:
        nodes, tets = cube_tet_mesh(n, side=side)

    K, M = assemble_global_p1(nodes, tets)
    mask = cube_interior_mask(nodes, side=side)
    K_int, M_int = apply_dirichlet(K, M, mask)
    eigvals, eigvecs = eigensolve(K_int, M_int, k=k)

    return {
        "n_nodes": int(nodes.shape[0]),
        "n_tets": int(tets.shape[0]),
        "n_int": int(K_int.shape[0]),
        "nodes": nodes,
        "tets": tets,
        "interior_mask": mask,
        "K": K,
        "M": M,
        "K_int": K_int,
        "M_int": M_int,
        "eigenvalues": eigvals,
        "eigenvectors": eigvecs,
        "k_frobenius": float(scipy.sparse.linalg.norm(K, "fro")),
        "m_frobenius": float(scipy.sparse.linalg.norm(M, "fro")),
        "k_int_frobenius": float(scipy.sparse.linalg.norm(K_int, "fro")),
        "m_int_frobenius": float(scipy.sparse.linalg.norm(M_int, "fro")),
        "k_int_diag": K_int.diagonal().astype(np.float64),
        "m_int_diag": M_int.diagonal().astype(np.float64),
    }


# --------------------------------------------------------------------------- #
# Analytic targets — Dirichlet eigenvalues of the unit cube Laplacian
# --------------------------------------------------------------------------- #


def analytic_lowest_five():
    """Lowest 5 Dirichlet Laplacian eigenvalues on the unit cube.

    Mode (p, q, r) has eigenvalue ``(p² + q² + r²)·π²``. The five
    smallest by ``p²+q²+r²``:

        (1,1,1) — 3π²
        (2,1,1), (1,2,1), (1,1,2) — 6π² (3-fold degenerate)
        (2,2,1) and permutations — 9π² (3-fold degenerate; we only
        pull the first into the lowest-5 set since the next 9π² mode
        is interleaved with other 9π² modes — see eigensolver tests
        ``crates/geode-core/examples/eigen_convergence.rs`` which uses
        exactly this 5-tuple for the same reason).
    """
    pi2 = np.pi**2
    return np.array(
        [3.0 * pi2, 6.0 * pi2, 6.0 * pi2, 6.0 * pi2, 9.0 * pi2], dtype=np.float64
    )


if __name__ == "__main__":
    # CLI: print the lowest-5 eigenvalues for the n=10 mesh.
    result = run_cube_cavity(n=10, k=5)
    targets = analytic_lowest_five()
    pi2 = np.pi**2

    print(f"NumPy cube-cavity n=10: n_int = {result['n_int']}")
    print()
    print("idx  target/π²   λ_h/π²    rel err")
    print("---  ---------   ------    ---------")
    for i, (got, want) in enumerate(zip(result["eigenvalues"], targets)):
        rel = abs(got - want) / want * 100.0
        print(f"{i:<3}  {want / pi2:.4f}      {got / pi2:.4f}    {rel:+.4f}%")

File: reference/julia/cube_cavity.jl
#!/usr/bin/env julia
"""
Julia reference for the scalar-Helmholtz cube-cavity eigenmode pipeline
(Epic #88 / Phase E, issue #115).

This is the **complex-arithmetic reference backend** per Epic #88's
principle 4. Even though the cube-cavity slice is real-symmetric,
landing Julia here lays the toolchain track for Phases G–J (Nédélec,
PML, NLEPS) where complex types matter, and surfaces complex-arithmetic
ergonomics friction earlier rather than later.

Algorithmic structure mirrors `reference/numpy/cube_cavity.py` exactly:

1. Mesh I/O — generate the canonical n-per-side tet-split cube via
   `CubeMesh.cube_tet_mesh(n)` **or** load a `.msh` via inline MSH 4.1
   parser (`CubeMesh.load_msh`). Both live in `mesh.jl`.
2. P1 local matrices — closed-form expressions transcribed faithfully
   from `reference/numpy/p1_local_matrices.py` (which itself is the
   reference for `crates/geode-core/src/p1.rs`).
3. Global assembly — COO triples → CSC via `SparseArrays.sparse(...)`
   which collapses duplicates by sum (same semantics as scipy's
   `coo_matrix.tocsr`).
4. Dirichlet BC — restrict K, M to interior nodes via fancy indexing.
5. Generalized eigensolve — `Arpack.eigs(K_int, M_int, nev=5, which=:SM)`.
   Arpack.jl binds the same `libarpack` as `scipy.sparse.linalg.eigsh`,
   making this a near-bit-equivalent of the NumPy reference at the
   iteration-trace level. The exact call differs from SciPy's
   `eigsh(K, k, M=M, sigma=0, which="LM")`: Arpack.jl 0.5 mis-handles
   that shift-invert recipe on the generalized pencil and returns the
   largest eigenvalues instead. See `eigensolve_arpack` for the gory
   details and the regular-inverse-mode workaround.

Usage
=====

Self-check (with the canonical n=10 mesh fixture):

    julia --project=. cube_cavity.jl

Programmatic invocation (called by gen_cube_cavity_baseline.jl and by
the CI gate):

    julia --project=. cube_cavity.jl --n 10 \\
        --mesh ../fixtures/cube_cavity/unit_cube.msh \\
        --out  ../fixtures/cube_cavity/julia_baseline.json

ARPACK iteration-trace caveat
=============================

Arpack.jl's `eigs` is sensitive to the Arnoldi starting vector. We
explicitly seed v0 (see `eigensolve_arpack`) so the run is
deterministic across machines. Eigenvalues are stable to ~1e-13
across re-runs; eigenvectors within a degenerate cluster require the
subspace-overlap convention (we do not store eigenvectors in
`julia_baseline.json` — mirroring `jax_baseline.json`'s shape).

Arpack.jl vs SciPy calling-convention divergence
================================================

Arpack.jl 0.5.x and SciPy share the same underlying libarpack, so they
agree to the iteration-trace level on the *same* operator. They do not
agree on how to set up the "lowest generalized eigenvalues" call. SciPy
uses `eigsh(K, k, M=M, sigma=0, which="LM")` (shift-invert at σ=0,
ask for largest-magnitude eigenvalues of `(K-σM)⁻¹M`); Arpack.jl 0.5
mis-routes that recipe through its `:auto` explicit-transform path and
returns the *largest* eigenvalues. We use `eigs(K, M; nev, which=:SM)`
(no `sigma`) — Arpack.jl's regular-inverse mode (`mode = 2`) — which
factorizes M once and Lanczos-iterates on `M⁻¹K`, asking for
smallest-magnitude. This matches the dense `eigvals(K, M)` result to
~1e-13 and is the Julia friction artifact for Epic #88 Phase E
(complex-arithmetic toolchain bring-up).
"""

push!(LOAD_PATH, @__DIR__)
include(joinpath(@__DIR__, "mesh.jl"))

using .CubeMesh: cube_tet_mesh, cube_interior_mask, load_msh
using LinearAlgebra
using SparseArrays
using Arpack
using JSON3
using Printf


# --------------------------------------------------------------------------- #
# P1 local matrices (per-tet)
# --------------------------------------------------------------------------- #

"""
    p1_local(verts) -> (k_local, m_local, signed_volume)

Compute P1 local stiffness and consistent mass for one tetrahedron.

Faithful transcription of `reference/numpy/p1_local_matrices.py` for a
single element. Named intermediates match the symbols in
`crates/geode-core/src/p1.rs` one-for-one: edges `e1 = v2-v1, e2 = v3-v1,
e3 = v4-v1`; area-weighted basis gradients `g1 = e2 x e3`, `g2 = e3 x
e1`, `g3 = e1 x e2`, `g0 = -(g1 + g2 + g3)`; `det = e1 . g1`; element
volume `V = |det| / 6`; `K_ij = (g_i . g_j) / (6 |det|)`; `M_ij = (V /
20)(1 + δ_ij)`.

# Arguments
- `verts::AbstractMatrix{Float64}` of shape `(4, 3)` — per-tet vertex
  coordinates.
"""
function p1_local(verts::AbstractMatrix{Float64})
    v0 = @view verts[1, :]
    v1 = @view verts[2, :]
    v2 = @view verts[3, :]
    v3 = @view verts[4, :]

    e1 = v1 .- v0
    e2 = v2 .- v0
    e3 = v3 .- v0

    g1 = cross(e2, e3)
    g2 = cross(e3, e1)
    g3 = cross(e1, e2)
    g0 = -(g1 .+ g2 .+ g3)

    det = dot(e1, g1)
    signed_volume = det / 6.0
    abs_det = abs(det)

    # Stack G as a 4x3 matrix with row i = g_{i-1}.
    G = Matrix{Float64}(undef, 4, 3)
    G[1, :] .= g0
    G[2, :] .= g1
    G[3, :] .= g2
    G[4, :] .= g3
    gg = G * transpose(G)  # 4x4 Gram matrix

    k_local = gg ./ (6.0 * abs_det)

    # Mass pattern: 2 on diagonal, 1 off-diagonal.
    m_local = abs_det / 120.0 .* Float64[
        2.0 1.0 1.0 1.0;
        1.0 2.0 1.0 1.0;
        1.0 1.0 2.0 1.0;
        1.0 1.0 1.0 2.0
    ]
    return k_local, m_local, signed_volume
end


# --------------------------------------------------------------------------- #
# Global assembly via SparseArrays.sparse(rows, cols, vals)
# --------------------------------------------------------------------------- #

"""
    assemble_global_p1(nodes, tets) -> (K, M)

Assemble global stiffness `K` and consistent mass `M` as
`SparseMatrixCSC{Float64}` matrices. Uses COO triples assembled into the
final CSC via `SparseArrays.sparse(I, J, V, m, n, +)` — the `+` combiner
collapses duplicate `(i, j)` entries by sum (scipy's `coo_matrix.tocsr`
default).

Mirrors `reference/numpy/cube_cavity.py::assemble_global_p1`.
"""
function assemble_global_p1(nodes::AbstractMatrix{Float64}, tets::AbstractMatrix{Int})
    n_nodes = size(nodes, 1)
    n_elem = size(tets, 1)

    # Reserve triplet storage: 16 entries per element.
    nnz_est = 16 * n_elem
    I = Vector{Int}(undef, nnz_est)
    J = Vector{Int}(undef, nnz_est)
    Vk = Vector{Float64}(undef, nnz_est)
    Vm = Vector{Float64}(undef, nnz_est)

    p = 1
    for e in 1:n_elem
        verts = nodes[tets[e, :], :]
        k_local, m_local, _ = p1_local(verts)
        for i in 1:4, j in 1:4
            I[p] = tets[e, i]
            J[p] = tets[e, j]
            Vk[p] = k_local[i, j]
            Vm[p] = m_local[i, j]
            p += 1
        end
    end

    K = sparse(I, J, Vk, n_nodes, n_nodes, +)
    M = sparse(I, J, Vm, n_nodes, n_nodes, +)
    return K, M
end


"""
    apply_dirichlet(K, M, mask) -> (K_int, M_int)

Restrict K, M to the rows/cols where `mask` is true. The dropped rows/
cols implement homogeneous Dirichlet conditions on the eliminated DOFs.
"""
function apply_dirichlet(K::SparseMatrixCSC{Float64}, M::SparseMatrixCSC{Float64},
                        mask::BitVector)
    idx = findall(mask)
    K_int = K[idx, idx]
    M_int = M[idx, idx]
    return K_int, M_int
end


# --------------------------------------------------------------------------- #
# Generalized eigensolve via Arpack.jl
# --------------------------------------------------------------------------- #

"""
    eigensolve_arpack(K, M; nev=5) -> (eigvals, eigvecs)

Lowest-`nev` generalized eigenpairs of `K x = λ M x` via Arpack.jl.

Arpack.jl wraps the same libarpack used by `scipy.sparse.linalg.eigsh`,
so the iteration trace agrees with the NumPy reference at the
bit-equivalent precision level on the same matrices.

**Calling convention — `which=:SM` (regular inverse mode), NOT
`sigma=0; which=:LM` (shift-invert).**

`scipy.sparse.linalg.eigsh(K, k, M=M, sigma=0.0, which="LM")` is the
canonical recipe for "lowest-`k` generalized eigenvalues" in SciPy.
**Arpack.jl 0.5 does not produce the same result on the generalized
problem with that call**: when `sigma !== nothing` and the problem is
generalized, Arpack.jl 0.5 internally takes the `:auto`
`explicittransform=:shiftinvert` path, swaps `:LM ↔ :SM`, factorizes
`σB - A = -K` (at σ=0), and ends up solving the standard problem for
`-K⁻¹ M` with `:SM`. The smallest-magnitude eigenvalues of `-K⁻¹ M`
correspond to the *largest* generalized eigenvalues of `K x = λ M x`,
so the post-processing step `λ = σ - 1/μ` returns the **largest**
generalized eigenvalues — the opposite of what the user requested.

For the generalized pencil with M SPD, Arpack.jl's "regular inverse
mode" (`mode = 2`, `which=:SM`, no `sigma`) does the right thing:
it factorizes M once, then runs Lanczos on the operator
`x ↦ M⁻¹ K x`, asking for the smallest-magnitude eigenvalues.

A deterministic v0 is supplied so repeated runs on the same machine
produce identical Arnoldi traces; eigenvalues are stable across
machines regardless.
"""
function eigensolve_arpack(K::SparseMatrixCSC{Float64}, M::SparseMatrixCSC{Float64};
                          nev::Int=5)
    n = size(K, 1)
    # Deterministic seed: all ones, normalized. Matches scipy's "supplied v0"
    # idiom; eliminates random-restart drift between machines.
    v0 = ones(Float64, n) ./ sqrt(n)
    # which=:SM in regular-inverse mode (no sigma) for the generalized
    # pencil — see the docstring above for why the SciPy-style
    # `sigma=0, which=:LM` call is buggy under Arpack.jl 0.5.
    eigvals, eigvecs = eigs(K, M; nev=nev, which=:SM, v0=v0)
    # Real symmetric pencil ⇒ imaginary parts are at round-off; strip them.
    eigvals_real = real.(eigvals)
    eigvecs_real = real.(eigvecs)
    # Sort ascending.
    order = sortperm(eigvals_real)
    return eigvals_real[order], eigvecs_real[:, order]
end


# --------------------------------------------------------------------------- #
# End-to-end driver
# --------------------------------------------------------------------------- #

"""
    run_cube_cavity(; n=10, k=5, mesh_path=nothing, side=1.0) -> NamedTuple

Run the full Julia cube-cavity pipeline. Either `n` (programmatic
mesh) or `mesh_path` (load a `.msh`) controls the mesh source.

Returns a `NamedTuple` with the same diagnostic fields as the NumPy
reference's `run_cube_cavity`:
- `n_nodes, n_tets, n_int`
- `nodes, tets`
- `interior_mask`
- `K, M, K_int, M_int`
- `eigenvalues, eigenvectors`
- `k_frobenius, m_frobenius, k_int_frobenius, m_int_frobenius`
- `k_int_diag, m_int_diag`
"""
function run_cube_cavity(; n::Int=10, k::Int=5,
                        mesh_path::Union{Nothing,AbstractString}=nothing,
                        side::Float64=1.0)
    nodes, tets = if mesh_path === nothing
        cube_tet_mesh(n; side=side)
    else
        load_msh(mesh_path)
    end

    K, M = assemble_global_p1(nodes, tets)
    mask = cube_interior_mask(nodes; side=side)
    K_int, M_int = apply_dirichlet(K, M, mask)
    eigvals, eigvecs = eigensolve_arpack(K_int, M_int; nev=k)

    return (
        n_nodes = size(nodes, 1),
        n_tets = size(tets, 1),
        n_int = size(K_int, 1),
        nodes = nodes,
        tets = tets,
        interior_mask = mask,
        K = K,
        M = M,
        K_int = K_int,
        M_int = M_int,
        eigenvalues = eigvals,
        eigenvectors = eigvecs,
        k_frobenius = norm(K),  # Frobenius for SparseMatrixCSC
        m_frobenius = norm(M),
        k_int_frobenius = norm(K_int),
        m_int_frobenius = norm(M_int),
        k_int_diag = Vector{Float64}(diag(K_int)),
        m_int_diag = Vector{Float64}(diag(M_int)),
    )
end


"""
    analytic_lowest_five() -> Vector{Float64}

Lowest 5 Dirichlet Laplacian eigenvalues on the unit cube:
`{3, 6, 6, 6, 9} * π²`.
"""
function analytic_lowest_five()
    π² = pi * pi
    return Float64[3.0 * π², 6.0 * π², 6.0 * π², 6.0 * π², 9.0 * π²]
end


# --------------------------------------------------------------------------- #
# CLI entry point
# --------------------------------------------------------------------------- #

function _parse_args(argv::Vector{String})
    n = 10
    k = 5
    side = 1.0
    mesh_path::Union{Nothing,String} = nothing
    out_path::Union{Nothing,String} = nothing
    i = 1
    while i <= length(argv)
        arg = argv[i]
        if arg == "--n"
            n = parse(Int, argv[i + 1]); i += 2
        elseif arg == "--k"
            k = parse(Int, argv[i + 1]); i += 2
        elseif arg == "--side"
            side = parse(Float64, argv[i + 1]); i += 2
        elseif arg == "--mesh"
            mesh_path = argv[i + 1]; i += 2
        elseif arg == "--out"
            out_path = argv[i + 1]; i += 2
        elseif arg in ("-h", "--help")
            println(stderr, """
            Usage: julia --project=. cube_cavity.jl [options]
                  --n <int>          cells per side (default 10)
                  --k <int>          number of eigenmodes (default 5)
                  --side <float>     cube edge length (default 1.0)
                  --mesh <path>      load mesh from .msh (overrides --n)
                  --out <path>       emit a sidecar JSON (used by the CI gate)
            """)
            exit(0)
        else
            error("unrecognized argument: $arg")
        end
    end
    return (n=n, k=k, side=side, mesh=mesh_path, out=out_path)
end


function _emit_sidecar(out_path::AbstractString, result::NamedTuple, k::Int)
    # Lightweight sidecar JSON in the same fixture-v1 shape that
    # `compare_eigenvalues.py` consumes (so the CI gate can read it
    # directly without language-specific code paths).
    n_recovered = _cbrt_int(result.n_tets ÷ 6)
    eigvals = result.eigenvalues[1:k]
    fixture = Dict(
        "schema_version" => "1",
        "fixture_id" => "cube_cavity/n$(n_recovered)_julia_eigensolve",
        "description" =>
            "Lowest $(k) cube-cavity eigenvalues from reference/julia/cube_cavity.jl. " *
            "Emitted in fixture-schema form so the cross-IR agreement table in " *
            "reference/driver/compare_eigenvalues.py can consume it without " *
            "language-specific code paths (Epic #88 / #115).",
        "units" => "dimensionless",
        "inputs" => Dict(
            "n_per_side" => Dict(
                "shape" => [1],
                "dtype" => "i64",
                "description" => "Hexes per cube side.",
                "data" => [n_recovered],
            ),
        ),
        "outputs" => Dict(
            "eigenvalues" => Dict(
                "shape" => [k],
                "dtype" => "f64",
                "description" => "Lowest $(k) eigenvalues, ascending.",
                "tolerance_abs" => 1.0e-8,
                "data" => collect(eigvals),
            ),
            "k_int_frobenius" => Dict(
                "shape" => [1],
                "dtype" => "f64",
                "description" => "Frobenius norm of K_int.",
                "tolerance_abs" => 1.0e-11,
                "data" => [result.k_int_frobenius],
            ),
            "m_int_frobenius" => Dict(
                "shape" => [1],
                "dtype" => "f64",
                "description" => "Frobenius norm of M_int.",
                "tolerance_abs" => 1.0e-13,
                "data" => [result.m_int_frobenius],
            ),
            "k_int_diag" => Dict(
                "shape" => [result.n_int],
                "dtype" => "f64",
                "description" => "Diagonal of K_int.",
                "tolerance_abs" => 1.0e-9,
                "data" => result.k_int_diag,
            ),
            "m_int_diag" => Dict(
                "shape" => [result.n_int],
                "dtype" => "f64",
                "description" => "Diagonal of M_int.",
                "tolerance_abs" => 5.0e-9,
                "data" => result.m_int_diag,
            ),
            "n_int" => Dict(
                "shape" => [1],
                "dtype" => "f64",
                "description" => "Number of interior DOFs.",
                "tolerance_abs" => 0.5,
                "data" => [Float64(result.n_int)],
            ),
        ),
        "provenance" => Dict(
            "source" => "reference/julia/cube_cavity.jl (Epic #88 / #115)",
            "julia_version" => string(VERSION),
            "issue" => "#115 (Phase E — Julia cube-cavity)",
        ),
    )
    open(out_path, "w") do io
        JSON3.pretty(io, fixture)
        write(io, "\n")
    end
    @info "Wrote sidecar JSON" path=out_path
end


# Round-trip cube root for small integers (used to recover `n` from
# `n_tets = 6 * n^3` in the sidecar JSON).
function _cbrt_int(n_cubed::Int)
    n = round(Int, cbrt(Float64(n_cubed)))
    n^3 == n_cubed || error("not a perfect cube: $n_cubed")
    return n
end


function main()
    args = _parse_args(ARGS)
    @info "Running Julia cube-cavity pipeline" n=args.n k=args.k side=args.side mesh=args.mesh
    result = run_cube_cavity(
        n=args.n, k=args.k, side=args.side, mesh_path=args.mesh
    )

    π² = pi * pi
    targets = analytic_lowest_five()
    println("Julia cube-cavity:")
    println("  n_nodes = $(result.n_nodes)")
    println("  n_tets  = $(result.n_tets)")
    println("  n_int   = $(result.n_int)")
    println()
    println("idx  target/π²   λ_h/π²    rel err")
    println("---  ---------   ------    ---------")
    for i in 1:length(result.eigenvalues)
        got = result.eigenvalues[i]
        if i <= length(targets)
            want = targets[i]
            rel = abs(got - want) / want * 100.0
            @printf("%-3d  %.4f      %.4f    %+.4f%%\n", i - 1, want / π², got / π², rel)
        else
            @printf("%-3d                %.4f\n", i - 1, got / π²)
        end
    end
    println()
    @printf("trace(K_int)       = %.12e\n", sum(result.k_int_diag))
    @printf("trace(M_int)       = %.12e\n", sum(result.m_int_diag))
    @printf("‖K_int‖_F          = %.12e\n", result.k_int_frobenius)
    @printf("‖M_int‖_F          = %.12e\n", result.m_int_frobenius)

    if args.out !== nothing
        _emit_sidecar(args.out, result, args.k)
    end
end


# Run only when invoked as a script (`julia --project=. cube_cavity.jl ...`),
# not when included from another module.
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
